File: backend/core/tonearm.py
import numpy as np


def calculate_servo_angle(
    groove_radius_mm: float,
    pivot_to_spindle_mm: float,
    arm_length_mm: float,
    servo_offset_deg: float = 0
):
    """
    Calculate servo angle
    using tonearm geometry.

    Triangle:

    pivot ↔ spindle = A
    pivot ↔ stylus  = B
    spindle ↔ groove = C
    """

    A = pivot_to_spindle_mm
    B = arm_length_mm
    C = groove_radius_mm

    cos_theta = (
        (A**2 + B**2 - C**2)
        /
        (2 * A * B)
    )

    # numerical safety
    cos_theta = np.clip(
        cos_theta,
        -1.0,
        1.0
    )

    theta_rad = np.arccos(
        cos_theta
    )

    theta_deg = np.degrees(
        theta_rad
    )

    return (
        theta_deg
        + servo_offset_deg
    )

File: wasm/src/pipeline/tonearm.ts
// Mirrors backend/core/tonearm.py → calculate_servo_angle()
// Law of cosines: A=pivot↔spindle, B=arm, C=groove_radius
// θ = arccos((A²+B²-C²)/(2AB)) + servo_offset

export const PIVOT_TO_SPINDLE_MM  = 220.0;
export const ARM_LENGTH_MM        = 195.0;
export const SERVO_ZERO_OFFSET_DEG = 90;

export function calculateServoAngle(
  grooveRadiusMm:     number,
  pivotToSpindleMm  = PIVOT_TO_SPINDLE_MM,
  armLengthMm       = ARM_LENGTH_MM,
  servoOffsetDeg    = SERVO_ZERO_OFFSET_DEG,
): number {
  const A = pivotToSpindleMm;
  const B = armLengthMm;
  const C = grooveRadiusMm;

  const cosTheta = Math.max(-1, Math.min(1, (A * A + B * B - C * C) / (2 * A * B)));
  const thetaDeg = (Math.acos(cosTheta) * 180) / Math.PI;

  return thetaDeg + servoOffsetDeg;
}
